### PR TITLE
Add loading of iconmap from configuration

### PR DESCRIPTION
**Pull request by fiskhest**
After setting up my environment to make use of the script, I figured if the repo was to ever be packaged or receive an update, a manual merge of my local changes would have to be performed for each update. I introduce this PR as a way of addressing that issue, since ICONS can now be defined in `~/.config/bspwm/bspi.iconmap` (or set the env variable `export bspi_config_path=/path/to/iconmap-file`) without having to modify the original script.

I am by no means a python expert, any suggestions for improvement is appreciated!

### CODE REVIEW

**Comment by aerosol** [comment]
Hey @fiskhest, this is really cool and thanks for taking time to do this.
I'm not a python expert either :) Would you mind considering the following:

1. Use https://docs.python.org/3/library/configparser.html for configuration format - possibly nicer than JSON?
2. Provide default configuration and make the script read it from an external file, placed by default in the script's location. I think this can be accomplished with:
```python
import os
os.path.dirname(os.path.realpath(__file__))
```
3. Instead of hardcoding a path, either fall back to [XDG](https://wiki.archlinux.org/index.php/XDG_Base_Directory) or add a [command line option](https://docs.python.org/3/library/argparse.html#module-argparse) with alternative config location (I think I like the latter better, what do you think?), e.g.
```
$ bspi.py --config $HOME/.config/bspi.conf
```

Happy to hear your thoughts. Thanks!

**Comment by fiskhest** [comment]
Hey @aerosol, thank you for the quick feedback and taking the time to create this! :-)

1. Absolutely. I was looking at it but took the shortcut of json since it didn't require too many changes to the original code. I'll rework it using ConfigParser.

2. Sure! So essentially, move the default `ICONS` dictionary to an external file?

3. Agreed! If it's not too hard, I'll work out a combination of the two, otherwise I'll add a command line option.

**Comment by aerosol** [comment]
> Sure! So essentially, move the default ICONS dictionary to an external file?

Yeah, if we make the defaults in a file that's read by default from script's own dir, whatever that is, we don't need the duplication I think. And by external file, we mean the INI-like `configparser` format right?

**Comment by fiskhest** [comment]
Yeah, sounds reasonable.

Yep, INI-formatted external file. Apologies for being implicit, good of you to make sure.

**Comment by fiskhest** [comment]
Modified and seems to work based on my simple testing.

Not entirely certain I got the xdg implementation correct. It's definitely working, but is my implementation something like what you had in mind?

File: bspi_iconmap.ini [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+signal = 
+slack = 
+spotify = 
+termite = 
+urxvt = 
+tilix = 
+alacritty = 
+kitty = 
+code-oss = 
+None = 
```
Comment (aerosol) [inline bspi_iconmap.ini]: Maybe since this is not pythonic `None` anymore, we could have a special catch-all key like `_other`?

File: bspi.py [hunk 2]
```diff
@@ -5,24 +5,40 @@
 import re
 import subprocess
 import sys
+import argparse
+import configparser
+import xdg
+
+script_dir = os.path.dirname(os.path.realpath(__file__))
+default_config_loc = f'{script_dir}/bspi_iconmap.ini'
```
Comment (aerosol) [inline bspi.py]: How about plain `bspi.ini`?

File: bspi.py [hunk 3]
```diff
@@ -5,24 +5,40 @@
...
+
+parser = argparse.ArgumentParser(description='bspi - rename bspwm desktops with icons based on the applications \
+    currently running under that desktop')
+parser.add_argument('-c', '--config', default=f'{default_config_loc}', help='absolute or relative path to the bspi \
```
Comment (aerosol) [inline bspi.py:16]: :ok_hand:

File: bspi.py [hunk 4]
```diff
@@ -5,24 +5,40 @@
...
-        return ICONS.get(self.client_class.lower(), ICONS[None])
+        return self.ICONS.get(self.client_class.lower(), self.ICONS.get('None'))
+
+    def _config(self):
+        config = configparser.ConfigParser()
+        if os.path.isfile(args.config):
+            cfg_file_path = args.config
+
+        else:
+            cfg_file_path = f'{xdg.XDG_CONFIG_HOME}/bspwm/bspi_iconmap.ini'
```
Comment (aerosol) [inline bspi.py]: Help suggests there is a default config path, that points to the stock configuration. 
Then `XDG_CONFIG_HOME` fallback works only if the config is explicitly provided *and* does not exist - that's a bit confusing.

I think we can give up on xdg dependency altogether, plain config overwrite via command line arg is IMHO enough. What do you think?

File: README.md [hunk 5]
```diff
@@ -10,10 +10,12 @@ FontAwesome icon set. Works with polybar.
...
 ## Prerequisites:
 
-bspwm (duh), xprop, python3, FontAwesome etc.
+bspwm (duh), xprop, python3(& pip), FontAwesome etc.
 
 ## Installation:
 
+Install requirements: `pip install --user -r requirements.txt`
+
 Run `bspi_listen &` in your `bspwmrc`.
```
Comment (aerosol) [inline README.md]: Ha, I forgot that our entry point is the `bspi_listen` script. Can we make it optionally accept the `--config` argument so it passes it to `bspi.py`? :thinking:

**Review by aerosol — COMMENTED** [review]
@fiskhest I just got my hands on my bspwm machine :-) It works, yay! 
I added some nitpicks inline. Let me know your thoughts.
Again, many thanks for your time. :heart:

File: README.md (hunk 5 above)
Comment (aerosol) [reply README.md]: `"$@"` should capture that I think

File: bspi.py (hunk 2 above)
Comment (fiskhest) [reply bspi.py]: absolutely. :-)

File: bspi.py (hunk 4 above)
Comment (fiskhest) [reply bspi.py]: Thanks for the feedback, it was precisely something like that I felt I had done, but didn't actually have time to untangle the logic when I was writing this. :fearful: Apologies!

Command line arg is definitely enough for this, and removes the external dependency on pip and the pip-package xdg. Much better :+1:

File: README.md (hunk 5 above)
Comment (fiskhest) [reply README.md]: Nice catch! I'd totally forgotten about it aswell :smile: . Yep, should be, I'll fix that!

**Comment by fiskhest** [comment]
Superb feedback, thank you!

**Comment by aerosol** [comment]
@fiskhest excellent work :tada:

**Comment by fiskhest** [comment]
Couldn't have done it without you, @aerosol :100: . Thanks! :+1:

**Pull request merged**